src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -0,0 +1,190 @@
+    private static final String ATTR_ACCOUNT_EXPIRES = "accountExpires";
+    // for Windows Server 2003-based domain
+    private static final String ATTR_USER_ACCOUNT_CONTROL_COMPUTED = "msDS-User-Account-Control-Computed";
+    // for ADAM (Active Directory Application Mode), replace the ADS_UF_DISABLED
+    private static final String ATTR_USER_ACCOUNT_DISABLED = "msDS-UserAccountDisabled";
+    // for ADAM, replace the ADS_UF_PASSWORD_EXPIRED
+    private static final String ATTR_USER_PASSWORD_EXPIRED = "msDS-UserPasswordExpired";
+
+    // https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // constant names follow the code in Iads.h
+    private static final long ACCOUNT_NO_EXPIRATION = 0x7FFF_FFFF_FFFF_FFFFL;
+    private static final int ADS_UF_DISABLED = 0x0002;
+    private static final int ADS_UF_LOCK_OUT = 0x0010;
+    private static final int ADS_UF_PASSWORD_EXPIRED = 0x80_0000;
+
+    public static boolean checkIfUserIsEnabled(@Nonnull Attributes user) {
+        try {
+            String userAccountControl = getStringAttribute(user, ATTR_USER_ACCOUNT_CONTROL);
+            if (userAccountControl != null) {
+                int uacAsInt = Integer.parseInt(userAccountControl);
+                if ((uacAsInt & ADS_UF_DISABLED) == ADS_UF_DISABLED) {
+                    return false;
+                }
+            }
+
+            String adamUserAccountDisabled = getStringAttribute(user, ATTR_USER_ACCOUNT_DISABLED);
+            if (adamUserAccountDisabled != null) {
+                if (adamUserAccountDisabled.equals("true")) {
+                    return false;
+                } else {
+                    return true;
+                }
+            }
+
+            return true;
+        } catch (NamingException e) {
+            return true;
+        }
+    }
+
+    public static boolean checkIfAccountNonExpired(@Nonnull Attributes user) {
+        try {
+            String accountExpirationDate = getStringAttribute(user, ATTR_ACCOUNT_EXPIRES);
+            if (accountExpirationDate != null) {
+                long expirationAsLong = Long.parseLong(accountExpirationDate);
+                if (expirationAsLong == 0L || expirationAsLong == ACCOUNT_NO_EXPIRATION) {
+                    return true;
+                }
+
+                long nowIn100NsFromJan1601 = getWin32EpochHundredNanos();
+                boolean expired = expirationAsLong < nowIn100NsFromJan1601;
+                return !expired;
+            }
+
+            return true;
+        } catch (NamingException e) {
+            return true;
+        }
+    }
+
+    // documentation: https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // code inspired by https://community.oracle.com/thread/1157460
+    private static long getWin32EpochHundredNanos() {
+        GregorianCalendar win32Epoch = new GregorianCalendar(1601, Calendar.JANUARY, 1);
```

Comment:
`ZonedDateTime.of(1601, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC)` could be easier to calculate with here.

Reply:
1) I followed the code as mentioned in the javadoc
2) I do not find the proposed one to be more readable than the one in the PR

If anybody else agree on your proposal I will change ;)

Reply:
My proposal was taking a random java.time class that sounded appropriate. Calendar is much more annoying to use.